apollo-runtime/src/main/java/com/apollographql/android/impl/RealApolloWatcher.java

```diff
@@ -0,0 +1,78 @@
+package com.apollographql.android.impl;
+
+import com.apollographql.android.ApolloCall;
+import com.apollographql.android.ApolloWatcher;
+import com.apollographql.android.api.graphql.Operation;
+import com.apollographql.android.api.graphql.Response;
+import com.apollographql.android.cache.normalized.Cache;
+import com.apollographql.android.cache.normalized.CacheControl;
+import com.apollographql.android.impl.util.Utils;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+final class RealApolloWatcher<T extends Operation.Data> implements ApolloWatcher<T> {
+
+  private RealApolloCall<T> activeCall;
+  @Nullable private ApolloCall.Callback<T> callback = null;
+  private final Cache cache;
+  private CacheControl refetchCacheControl = CacheControl.CACHE_FIRST;
+  private volatile boolean isSubscribed = true;
+  private final Cache.RecordChangeSubscriber recordChangeSubscriber = new Cache.RecordChangeSubscriber() {
+    @Override public void onDependentKeysChanged() {
+      refetch();
+    }
+  };
+  private final WatcherSubscription watcherSubscription = new WatcherSubscription() {
+    @Override public void unsubscribe() {
+      isSubscribed = false;
+      cache.unsubscribe(recordChangeSubscriber);
+    }
+  };
+
+  public RealApolloWatcher(RealApolloCall<T> originalCall, Cache cache) {
+    activeCall = originalCall;
+    this.cache = cache;
+  }
+
+  @Nonnull public WatcherSubscription enqueueAndWatch(@Nullable final ApolloCall.Callback<T> callback) {
+    this.callback = callback;
```

Comment:
Should we add safe guard if we already watching? I mean we don't support multiple subscription right?